# What is NumPy?
- It is multi-dimensional array libray.
- NumPy is a Python library.
- NumPy is used for working with arrays.
- NumPy is short for "Numerical Python".
- NumPy is a Python library used for working with arrays.
- It also has functions for working in domain of linear algebra, fourier transform, and matrices.
- NumPy was created in 2005 by Travis Oliphant. It is an open source project and you can use it freely.

# Why use NumPy?
- In Python we have lists that serve the purpose of arrays, but they are slow to process.
- NumPy aims to provide an array object that is up to 50x faster than traditional Python lists.
- The array object in NumPy is called ndarray, it provides a lot of supporting functions that make working with ndarray very easy.
- Arrays are very frequently used in data science, where speed and resources are very important.

# How are lists different from NumPy?
- Lists are slow
- NumPy is very fast

# Why NumPy is faster
- Faster to read less bytes of memory .
- No type checking when iterating through objects.
- It has fixed type
- It uses contiguous memory
- NumPy arrays are stored at one continuous place in memory unlike lists, so processes can access and manipulate them very efficiently.
- This behavior is called locality of reference in computer science.
- This is the main reason why NumPy is faster than lists. Also it is optimized to work with latest CPU architectures.
- NumPy is a Python library and is written partially in Python, but most of the parts that require fast computation are written in C or C++.

# Applications of NumPy Library
1. Matlab Replacement
2. Plotting
3. Machine Learning

In [2]:
import numpy as np

In [3]:
a = np.array([1,2,3])
print(a)

[1 2 3]


In [4]:
b = np.array([[9.0,8.0,7.0],[6.0,5.0,4.0]])
print(b)

[[9. 8. 7.]
 [6. 5. 4.]]


In [6]:
# Get Dimensions of your array
a.ndim

1

In [7]:
b.ndim

2

In [8]:
# Get Shape
a.shape

(3,)

In [10]:
b.shape # because it has two rows and three columns

(2, 3)

In [11]:
# Get Type
a.dtype

dtype('int64')

In [12]:
b.dtype

dtype('float64')

In [14]:
# Get Size
a.itemsize

8

In [15]:
b.itemsize

8

In [16]:
# Get total size
a.nbytes

24

In [17]:
# Get number of elements
a.size

3

# Accessing/Changing specific elements, rows, columns, etc

In [19]:
a = np.array([[1,2,3,4,5,6,7],[8,9,10,11,12,13,14]])
print(a)

[[ 1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14]]


In [20]:
a.shape

(2, 7)

In [25]:
# Get a specific element [row,column]
print(a[1,2])

10


In [29]:
# Get a specific row
print(a[0,:])

[1 2 3 4 5 6 7]


In [32]:
# Get a specific column
print(a[:,2])

[ 3 10]


In [39]:
# Getting a little more fancy [startindex:endindex:stepsize]
a[0,1:6:2]

array([2, 4, 6])

In [40]:
a[1,5] = 20

In [41]:
a

array([[ 1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 20, 14]])

In [44]:
a[:,2] = 5

In [45]:
a

array([[ 1,  2,  5,  4,  5,  6,  7],
       [ 8,  9,  5, 11, 12, 20, 14]])

# 3-D EXAMPLE

In [51]:
b = np.array([[[1,2],[3,4]],[[5,6],[7,8]]])
print(b)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


In [53]:
# Get specific element 
print(b[1,0,1])

6


In [54]:
print(b[0,1,1])

4


In [56]:
print(b[:,1,:])

[[3 4]
 [7 8]]


In [57]:
print(b[:,0,:])

[[1 2]
 [5 6]]


In [59]:
print(b[:,0,0])

[1 5]


In [64]:
b[:,1,:] = [[9,9],[8,8]]
print(b)

[[[1 2]
  [9 9]]

 [[5 6]
  [8 8]]]


# Initializing Different Types of Arrays

In [66]:
# All 0s matrix
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [70]:
# All 1s matrix
np.ones((3,8),dtype='int')

array([[1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1]])

In [73]:
# Any other number np.full((shape),number)
np.full((4,4),99,dtype='float32')

array([[99., 99., 99., 99.],
       [99., 99., 99., 99.],
       [99., 99., 99., 99.],
       [99., 99., 99., 99.]], dtype=float32)

In [80]:
# Any other number (full_like)
np.full_like(a,3)

array([[3, 3, 3, 3, 3, 3, 3],
       [3, 3, 3, 3, 3, 3, 3]])

In [83]:
# Random decimal numbers
np.random.rand(2,2)

array([[0.11929663, 0.74446419],
       [0.71006001, 0.03321281]])

In [91]:
# Random decimal numbers
np.random.random_sample(a.shape)

array([[0.36712038, 0.63211983, 0.41828581, 0.38505635, 0.19533625,
        0.33782598, 0.96058552],
       [0.11796017, 0.83720332, 0.98375944, 0.99757106, 0.62634586,
        0.42133505, 0.59691908]])

In [149]:
# Random integer numbers
np.random.randint(-5,6,size=(3,3))

array([[-5, -3,  2],
       [ 3, -5,  2],
       [-3,  0,  1]], dtype=int32)

In [154]:
# The identity matrix
np.identity(2,dtype='int64')

array([[1, 0],
       [0, 1]])

In [165]:
# Repeat an array
arr = np.array([[1,2,3]])
r1 = np.repeat(arr,3,axis=0)
print(r1)

[[1 2 3]
 [1 2 3]
 [1 2 3]]


In [180]:
output = np.ones((5,5))
#print(output)

z = np.zeros((3,3),dtype='int')
z[1,1] = 9
#print(z)
output[1:4,1:4] = z
print(output)

[[1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 1.]
 [1. 0. 9. 0. 1.]
 [1. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1.]]


In [8]:
a = np.array([1,2,3])
#print(a)
b = a.copy()
#print(b)
b[0] = 100
print(a)
print(b)

[1 2 3]
[100   2   3]


# Mathematics

In [9]:
a = np.array([1,2,3,4])
print(a)

[1 2 3 4]


In [10]:
a + 2

array([3, 4, 5, 6])

In [11]:
a - 2

array([-1,  0,  1,  2])

In [12]:
a * 2

array([2, 4, 6, 8])

In [13]:
a / 2

array([0.5, 1. , 1.5, 2. ])

In [15]:
a = a + 2
print(a)

[5 6 7 8]


In [16]:
a ** 2

array([25, 36, 49, 64])

In [18]:
# Take the sine
np.sin(a)

array([-0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825])

In [19]:
# Take the cosine
np.cos(a)

array([ 0.28366219,  0.96017029,  0.75390225, -0.14550003])

#### Linear Algebra

In [26]:
a = np.ones((2,3))
print(a)
b = np.full((3,2),2)
print(b)

np.matmul(a,b)

[[1. 1. 1.]
 [1. 1. 1.]]
[[2 2]
 [2 2]
 [2 2]]


array([[6., 6.],
       [6., 6.]])

In [29]:
# Find the determinant
c = np.identity(3)
np.linalg.det(c)

np.float64(1.0)

In [33]:
# Version of NumPy
print(np.__version__)

2.1.0


#### Statistics

In [4]:
stats = np.array([[1,2,3],[4,5,6]])
stats

array([[1, 2, 3],
       [4, 5, 6]])

In [7]:
np.min(stats)

np.int64(1)

In [8]:
np.max(stats)

np.int64(6)

In [9]:
np.mean(stats)

np.float64(3.5)

In [10]:
np.median(stats)

np.float64(3.5)

In [11]:
np.std(stats)

np.float64(1.707825127659933)

In [12]:
np.sum(stats)

np.int64(21)

# Reorganizing arrays

In [24]:
before = np.array([[1,2,3,4],[5,6,7,8]])
print(before)
print(before.shape)

[[1 2 3 4]
 [5 6 7 8]]
(2, 4)


In [25]:
after = before.reshape((4,2))
print(after)

[[1 2]
 [3 4]
 [5 6]
 [7 8]]


In [26]:
print(after.reshape((2,4)))

[[1 2 3 4]
 [5 6 7 8]]


In [27]:
print(after.reshape((1,8)))

[[1 2 3 4 5 6 7 8]]


In [28]:
print(after.reshape((8,1)))

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]
 [8]]


In [31]:
# Vertically Stacking Vectors
n1 = np.array([1,2,3,4])
n2 = np.array([5,6,7,8])
np.vstack([n1,n2,n1,n2])

array([[1, 2, 3, 4],
       [5, 6, 7, 8],
       [1, 2, 3, 4],
       [5, 6, 7, 8]])

In [38]:
h1 = np.zeros((2,2))
h2 = np.ones((2,2))
np.hstack([h1,h2])

array([[0., 0., 1., 1.],
       [0., 0., 1., 1.]])

# Miscellaneous

#### Load Data From File

In [57]:
filedata = np.genfromtxt('data.txt',delimiter=',')
filedata = filedata.astype('int32')
filedata

array([[  1,  13,  21,  11, 196,  75,   4,   3,  34,   6,   7,   8,   0,
          1,   2,   3,   4,   5],
       [  3,  42,  12,  33, 766,  75,   4,  55,   6,   4,   3,   4,   5,
          6,   7,   0,  11,  12],
       [  1,  22,  33,  11, 999,  11,   2,   1,  78,   0,   1,   2,   9,
          8,   7,   1,  76,  88]], dtype=int32)

### Boolean Masking and Advanced Indexing

In [63]:
filedata[filedata > 50]

array([196,  75, 766,  75,  55, 999,  78,  76,  88], dtype=int32)

In [68]:
# You can index with a list in NumPy
a = np.array([1,2,3,4,5,6,7,8,9])
a[[1,2,8]]

array([2, 3, 9])

In [69]:
np.any(filedata > 50,axis=0)

array([False, False, False, False,  True,  True, False,  True,  True,
       False, False, False, False, False, False, False,  True,  True])

In [70]:
np.all(filedata > 50,axis=0)

array([False, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False])

In [72]:
((filedata > 50) & (filedata <100))

array([[False, False, False, False, False,  True, False, False, False,
        False, False, False, False, False, False, False, False, False],
       [False, False, False, False, False,  True, False,  True, False,
        False, False, False, False, False, False, False, False, False],
       [False, False, False, False, False, False, False, False,  True,
        False, False, False, False, False, False, False,  True,  True]])

In [74]:
(~((filedata > 50) & (filedata < 100)))

array([[ True,  True,  True,  True,  True, False,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True, False,  True, False,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True,  True, False,
         True,  True,  True,  True,  True,  True,  True, False, False]])

## Synopsis

###### 0-D Arrays
- 0-D arrays, or Scalars, are the elements in an array. Each value in an array is a 0-D array.
###### 1-D Arrays
- An array that has 0-D arrays as its elements is called uni-dimensional or 1-D array.
###### 2-D Arrays
- An array that has 1-D arrays as its elements is called a 2-D array.
- These are often used to represent matrix or 2nd order tensors.
###### 3-D arrays
- An array that has 2-D arrays (matrices) as its elements is called 3-D array.
- These are often used to represent a 3rd order tensor.
- NumPy Arrays provides the ndim attribute that returns an integer that tells us how many dimensions the array have.

###### Access Array Elements
- Array indexing is the same as accessing an array element.
- You can access an array element by referring to its index number.
- The indexes in NumPy arrays start with 0, meaning that the first element has index 0, and the second has index 1 etc.
###### Access 2-D Arrays
- To access elements from 2-D arrays we can use comma separated integers representing the dimension and the index of the element.
- Think of 2-D arrays like a table with rows and columns, where the dimension represents the row and the index represents the column.
###### Access 3-D Arrays
- To access elements from 3-D arrays we can use comma separated integers representing the dimensions and the index of the element.
- Use negative indexing to access an array from the end.

###### Slicing arrays
- Slicing in python means taking elements from one given index to another given index.
- We pass slice instead of index like this: [start:end].
- We can also define the step, like this: [start:end:step].
- If we don't pass start its considered 0
- If we don't pass end its considered length of array in that dimension
- If we don't pass step its considered 1.
- `Note:` The result includes the start index, but excludes the end index.
- Use the minus operator to refer to an index from the end.
- Use the step value to determine the step of the slicing.

###### Data Types in Python
- By default Python have these data types:
- strings - used to represent text data, the text is given under quote marks. e.g. "ABCD"
- integer - used to represent integer numbers. e.g. -1, -2, -3
- float - used to represent real numbers. e.g. 1.2, 42.42
- boolean - used to represent True or False.
- complex - used to represent complex numbers. e.g. 1.0 + 2.0j, 1.5 + 2.5j
###### Data Types in NumPy
- NumPy has some extra data types, and refer to data types with one character, like i for integers, u for unsigned integers etc.
- Below is a list of all data types in NumPy and the characters used to represent them.
i - integer
b - boolean
u - unsigned integer
f - float
c - complex float
m - timedelta
M - datetime
O - object
S - string
U - unicode string
V - fixed chunk of memory for other type ( void )
- Checking the Data Type of an Array
- The NumPy array object has a property called dtype that returns the data type of the array.

###### Converting Data Type on Existing Arrays
- The best way to change the data type of an existing array, is to make a copy of the array with the astype() method.
- The astype() function creates a copy of the array, and allows you to specify the data type as a parameter.
- The data type can be specified using a string, like 'f' for float, 'i' for integer etc. or you can use the data type directly like float for float and int for integer.

###### The Difference Between Copy and View
- The main difference between a copy and a view of an array is that the copy is a new array, and the view is just a view of the original array.
- The copy owns the data and any changes made to the copy will not affect original array, and any changes made to the original array will not affect the copy.
- The view does not own the data and any changes made to the view will affect the original array, and any changes made to the original array will affect the view.

###### Check if Array Owns its Data
- As mentioned above, copies owns the data, and views does not own the data, but how can we check this?
- Every NumPy array has the attribute base that returns None if the array owns the data.
- Otherwise, the base  attribute refers to the original object.
- The copy returns None.
- The view returns the original array.

###### Reshaping arrays
- Reshaping means changing the shape of an array.
- The shape of an array is the number of elements in each dimension.
- By reshaping we can add or remove dimensions or change number of elements in each dimension.
###### Flattening the arrays
- Flattening array means converting a multidimensional array into a 1D array.
- We can use reshape(-1) to do this.